add backup configuration document (#1754)
vdaas/vald: docs/user-guides/backup-configuration.md

# Backup configuration

There are three types of options for the Vald cluster: backup, filtering, and core algorithms.

> This page describes how enabled backup features for your Vald cluster.

Reviewer: ```suggestion
This page describes how to enable backup feature on your Vald cluster.
```

## What is the backup

> Vald's backup function is to save the index information indexed in each Vald Agent pod as file data in Persistent Volume or S3.

Reviewer: ```suggestion
Vald's backup function is to save the index data in each Vald Agent pod as a data file to the Persistent Volume or S3.
```

When the Vald Agent pod is restarted for some reason, the index state is restored from the saved index data.

## Backup configuration

> This section shows the best practice backup configuration with PV, S3, or PV + S3.

Reviewer: ```suggestion
This section shows the best practice to configure backup features with PV, S3, or PV + S3.
```

Each sample configuration yaml is published on [here](https://github.com/vdaas/vald/tree/master/charts/vald/values).

> Please try to reference.

Reviewer: ```suggestion
```

Author: ```suggestion
Please try to reference it for more details.
```

Reviewer: ```suggestion
Please refer it for more details.
```

### General

Regardless of the backup destination, the following Vald Agent settings must be set to enable backup.

> - `index_path` is the location where those files are stored.

Reviewer: ```suggestion
- `index_path` is the backup file location.
```

  terminationGracePeriodSeconds: 3600
  ngt:
    ...
    index_path: "/var/ngt/index"
    enable_in_memory_mode: false
    ...
```

### Persistent Volume

#### requirement

You must prepare PV before deployment when using Kubernetes Persistent Volume (PV) for backup storage.
Please refer to the setup guide of the usage environment for the provisioning PV.

For example:
- [GKE setup PV document](https://cloud.google.com/kubernetes-engine/docs/concepts/persistent-volumes)
- [EKS storage document](https://docs.aws.amazon.com/eks/latest/userguide/storage.html)

#### configuration

After provisioning PV, the following parameters are needed to be set.

```yaml
agent:
  ...
  persistentVolume:
    # use PV flag
    enabled: true
    # accessMode for PV (please verify your environment)
    accessMode: ReadWriteOncePod
    # storage class for PV (please verify your environment)
    storageClass: local-path
    # set enough size for backup
    size: 2Gi
  ...
```

Each PV will be mounted on each Vald Agent Pod's `index_path`.

It is highly recommended to set `copy_on_write` (CoW) as `true`. 

<div class="notice">
The CoW is an option to update the backup file safely.<BR>

> The backup file may be corrupted, and the Vald Agent pod may not restore from backup files when the Vald Agent pod terminates during saveIndex without CoW is not be enabled.<BR>

Reviewer: ```suggestion
The backup file may be corrupted, and the Vald Agent pod may not be able to restore from backup files when the Vald Agent pod terminates during saveIndex without CoW is not be enabled.<BR>
```

Author: ```suggestion
The backup file may be corrupted, and the Vald Agent pod may not be able to restore from backup files when the Vald Agent pod terminates during the save index function without CoW is not be enabled.<BR>
```

| :---------- | :---------------------------------------------------------------------------------------------------------------------------------------------------------- | :---------------------------------------------------------------------------------------------------- | :----------------------------------------------------------------------------------------------------------------------------------------------- |
| usecase     | Want to use backup with low cost<BR>Would not like to use some external storage for backup<BR>Want to backup with highly compatible storage with Kubernetes | Want to use the same backup file with several Vald clusters<BR>Want to access the backup files easily | Want to use backup with PV basically and access the backup files easily<BR>Want to prevent backup file failure due to Kubernetes cluster failure |

> | merit :+1:  | Easy to use<BR>Highly compatible with Kubernetes<BR>Complete with internal network<BR>Safety backup if applying CoW                                         | Easy to access backup files<BR>It fan be shared and used by multiple clusters                         | The safest of these methods                                                                                                                      |

Reviewer 2: ```suggestion
| merit :+1:  | Easy to use<BR>Highly compatible with Kubernetes<BR>Complete with internal network<BR>Safety backup if applying CoW                                         | Easy to access backup files<BR>It can be shared and used by multiple clusters                         | The safest of these methods                                                                                                                      |
```

> | demerit :x: | A bit hard to check backup files<BR>Can not share backup files for several Vald clusters                                                                    | Need to communicate with external network                                                             | Need to operate both storages<BR>The most expensive way                                                                                          |

Reviewer 2: ```suggestion
| cons :-1: | A bit hard to check backup files<BR>Can not share backup files for several Vald clusters                                                                    | Need to communicate with external network                                                             | Need to operate both storages<BR>The most expensive way                                                                                          |
```

In addition, `agent.terminationGracePeriodSeconds` value should be long enough to ensure the backup speed.

```yaml
agent:
  ...
  # We recommend setting this value long enough to ensure the backup speed of PV, since the Index is backed up at the end of the pod.
  terminationGracePeriodSeconds: 3600
  ngt:
    ...
    index_path: "/var/ngt/index"
    enable_in_memory_mode: false
    ...
```

### Persistent Volume

#### requirement

You must prepare PV before deployment when using Kubernetes Persistent Volume (PV) for backup storage.
Please refer to the setup guide of the usage environment for the provisioning PV.

For example:

- [GKE setup PV document](https://cloud.google.com/kubernetes-engine/docs/concepts/persistent-volumes)
- [EKS storage document](https://docs.aws.amazon.com/eks/latest/userguide/storage.html)

#### configuration

After provisioning PV, the following parameters are needed to be set.

```yaml
agent:
  ...
  persistentVolume:
    # use PV flag
    enabled: true
    # accessMode for PV (please verify your environment)
    accessMode: ReadWriteOncePod
    # storage class for PV (please verify your environment)

>     storageClass: local-path

Reviewer 2: I think it would be better to set GKE storage class and configuration as a sample.

| :-------- | :---------------------------------------------------------------------------------------------------------------------------------------------------------- | :---------------------------------------------------------------------------------------------------- | :----------------------------------------------------------------------------------------------------------------------------------------------- |
| usecase   | Want to use backup with low cost<BR>Would not like to use some external storage for backup<BR>Want to backup with highly compatible storage with Kubernetes | Want to use the same backup file with several Vald clusters<BR>Want to access the backup files easily | Want to use backup with PV basically and access the backup files easily<BR>Want to prevent backup file failure due to Kubernetes cluster failure |

> | pros :+1: | Easy to use<BR>Highly compatible with Kubernetes<BR>Low latency using in-cluster network<BR>Safety backup using Copy on Write option                        | Easy to access backup files<BR>It can be shared and used by multiple clusters                         | The safest of these methods                                                                                                                      |

Reviewer: About the pros of PV `It can be shared and used by multiple clusters`, 
does it means the same PV backup can be shared and used by multiple instance of vald agent? 
If so, will it conflict with different agent instances running at the same time when performing CRUD?

Author: You may misread :thinking:

Reviewer: Yes, sorry it was my misunderstanding.
